### PR TITLE
Fix formats parser resetting, test processing bad messages in kafka

### LINKED ISSUE #36274: Segfault (Null pointer dereference) ingesting from Kafka ProtobufSingle
**Describe what's wrong**

Segfault during ingest from Kafka (ProtobufSingle). 

**Does it reproduce on recent release?**

I first saw it on 22.1.3.7. Upgrading to 22.3.3.44 doesn't help. 

```
2022.04.14 17:40:22.243813 [ 54703 ] {} <Fatal> BaseDaemon: ########################################
2022.04.14 17:40:22.243859 [ 54703 ] {} <Fatal> BaseDaemon: (version 22.3.3.44 (official build), build id: F9D3C2B8666BEF5D) (from thread 47161) (no query) Received signal Segmentation fault (11
)
2022.04.14 17:40:22.243885 [ 54703 ] {} <Fatal> BaseDaemon: Address: NULL pointer. Access: read. Address not mapped to object.
2022.04.14 17:40:22.243908 [ 54703 ] {} <Fatal> BaseDaemon: Stack trace: 0x16a6c5e8 0x16a6541b 0x16a0766d 0x1696c3f0 0x168fc2b5 0x168fbe7a 0x16925ee2 0x152fb4ef 0x152fdb6f 0x168fc2b5 0x168fbe7a
0x16b53782 0x1691c603 0x1691005e 0x1690f3ab 0x1690ec78 0x1690d724 0x152f0f0a 0x152eedf8 0x14f02cee 0x14f05667 0x14f06670 0xb418757 0xb41c2dd 0x7fef823e16db 0x7fef8210a61f
2022.04.14 17:40:22.243988 [ 54703 ] {} <Fatal> BaseDaemon: 2. DB::ProtobufReader::endMessage(bool) @ 0x16a6c5e8 in /usr/bin/clickhouse
2022.04.14 17:40:22.244005 [ 54703 ] {} <Fatal> BaseDaemon: 3. ? @ 0x16a6541b in /usr/bin/clickhouse
2022.04.14 17:40:22.244035 [ 54703 ] {} <Fatal> BaseDaemon: 4. DB::ProtobufRowInputFormat::readRow(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, DB::RowReadExtension&) @ 0x16a0766d in /usr/bin/clickhouse
2022.04.14 17:40:22.244054 [ 54703 ] {} <Fatal> BaseDaemon: 5. DB::IRowInputFormat::generate() @ 0x1696c3f0 in /usr/bin/clickhouse
2022.04.14 17:40:22.244074 [ 54703 ] {} <Fatal> BaseDaemon: 6. DB::ISource::tryGenerate() @ 0x168fc2b5 in /usr/bin/clickhouse
2022.04.14 17:40:22.244090 [ 54703 ] {} <Fatal> BaseDaemon: 7. DB::ISource::work() @ 0x168fbe7a in /usr/bin/clickhouse
2022.04.14 17:40:22.244103 [ 54703 ] {} <Fatal> BaseDaemon: 8. DB::StreamingFormatExecutor::execute() @ 0x16925ee2 in /usr/bin/clickhouse
2022.04.14 17:40:22.244131 [ 54703 ] {} <Fatal> BaseDaemon: 9. DB::KafkaSource::generateImpl() @ 0x152fb4ef in /usr/bin/clickhouse
2022.04.14 17:40:22.244143 [ 54703 ] {} <Fatal> BaseDaemon: 10. DB::KafkaSource::generate() @ 0x152fdb6f in /usr/bin/clickhouse
2022.04.14 17:40:22.244155 [ 54703 ] {} <Fatal> BaseDaemon: 11. DB::ISource::tryGenerate() @ 0x168fc2b5 in /usr/bin/clickhouse
2022.04.14 17:40:22.244167 [ 54703 ] {} <Fatal> BaseDaemon: 12. DB::ISource::work() @ 0x168fbe7a in /usr/bin/clickhouse
2022.04.14 17:40:22.244190 [ 54703 ] {} <Fatal> BaseDaemon: 13. DB::SourceWithProgress::work() @ 0x16b53782 in /usr/bin/clickhouse
2022.04.14 17:40:22.244204 [ 54703 ] {} <Fatal> BaseDaemon: 14. DB::ExecutionThreadContext::executeTask() @ 0x1691c603 in /usr/bin/clickhouse
2022.04.14 17:40:22.244219 [ 54703 ] {} <Fatal> BaseDaemon: 15. DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x1691005e in /usr/bin/clickhouse
2022.04.14 17:40:22.244231 [ 54703 ] {} <Fatal> BaseDaemon: 16. DB::PipelineExecutor::executeImpl(unsigned long) @ 0x1690f3ab in /usr/bin/clickhouse
2022.04.14 17:40:22.244244 [ 54703 ] {} <Fatal> BaseDaemon: 17. DB::PipelineExecutor::execute(unsigned long) @ 0x1690ec78 in /usr/bin/clickhouse
2022.04.14 17:40:22.244256 [ 54703 ] {} <Fatal> BaseDaemon: 18. DB::CompletedPipelineExecutor::execute() @ 0x1690d724 in /usr/bin/clickhouse
2022.04.14 17:40:22.244269 [ 54703 ] {} <Fatal> BaseDaemon: 19. DB::StorageKafka::streamToViews() @ 0x152f0f0a in /usr/bin/clickhouse
2022.04.14 17:40:22.244281 [ 54703 ] {} <Fatal> BaseDaemon: 20. DB::StorageKafka::threadFunc(unsigned long) @ 0x152eedf8 in /usr/bin/clickhouse
2022.04.14 17:40:22.244302 [ 54703 ] {} <Fatal> BaseDaemon: 21. DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x14f02cee in /usr/bin/clickhouse
2022.04.14 17:40:22.244314 [ 54703 ] {} <Fatal> BaseDaemon: 22. DB::BackgroundSchedulePool::threadFunction() @ 0x14f05667 in /usr/bin/clickhouse
2022.04.14 17:40:22.244326 [ 54703 ] {} <Fatal> BaseDaemon: 23. ? @ 0x14f06670 in /usr/bin/clickhouse
2022.04.14 17:40:22.244359 [ 54703 ] {} <Fatal> BaseDaemon: 24. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xb418757 in /usr/bin/clickhouse
2022.04.14 17:40:22.244371 [ 54703 ] {} <Fatal> BaseDaemon: 25. ? @ 0xb41c2dd in /usr/bin/clickhouse
2022.04.14 17:40:22.244403 [ 54703 ] {} <Fatal> BaseDaemon: 26. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
2022.04.14 17:40:22.244427 [ 54703 ] {} <Fatal> BaseDaemon: 27. __clone @ 0x12161f in /lib/x86_64-linux-gnu/libc-2.27.so
```

My Kafka table is as follow. And there's an MV pulling data from it.
```
create table if not exists kafka_table (
  `mid`              String,
  `arid`                     String,
  `decision_source`                     UInt8,
  `targets`  Nested (
    `uid`                    UInt64,
    `target_type`                       UInt64,
    `name`              Nullable(String),
    `blurb`            Nullable(String),
    `name2`          Nullable(String),
    `col`    Array(Nullable(String))
  ),
  `identifier`                String,
  `note`                      String,
  `created`             DateTime,
  `is_final`                   UInt8
)
ENGINE = Kafka
SETTINGS kafka_broker_list = '...',
       kafka_topic_list = '...',
       kafka_group_name = 'clickhouse',
       kafka_format = 'ProtobufSingle',
       kafka_schema = '...',
       kafka_handle_error_mode='stream'
;
```

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix formats parser resetting, test processing bad messages in kafka. Closes https://github.com/ClickHouse/ClickHouse/issues/36274

### CODE REVIEW

File: src/Processors/Executors/StreamingFormatExecutor.cpp [hunk 1]
```diff
@@ -93,6 +93,12 @@ size_t StreamingFormatExecutor::execute()
         format->resetParser();
         return on_error(result_columns, e);
     }
+    catch (std::exception & e)
```
Comment (kssenii) [inline src/Processors/Executors/StreamingFormatExecutor.cpp:96]: why std exception is enough?

Comment (Avogar) [reply src/Processors/Executors/StreamingFormatExecutor.cpp:96]: It's not, thanks for noticing. I will add processing any exception.

Comment (Avogar) [reply src/Processors/Executors/StreamingFormatExecutor.cpp:97]: Fixed

**Review by kssenii — APPROVED** [review]

File: tests/integration/test_storage_kafka/test.py [hunk 2]
```diff
@@ -4440,6 +4440,171 @@ def test_block_based_formats_2(kafka_cluster):
...
+                SELECT _error FROM test.kafka WHERE length(_error) != 0 ;                
+        """
+    )
+
+    print("CapnProto")
+
+    kafka_create_topic(admin_client, "CapnProto_err")
+
+    messages = ["qwertyuiop", "asdfghjkl", "zxcvbnm"]
+    kafka_produce(kafka_cluster, 'CapnProto_err', messages)
```
Comment (Avogar) [inline tests/integration/test_storage_kafka/test.py:4593]:
```suggestion
    kafka_produce(kafka_cluster, "CapnProto_err", messages)
```

File: tests/integration/test_storage_kafka/test.py [hunk 3]
```diff
@@ -4440,6 +4440,171 @@ def test_block_based_formats_2(kafka_cluster):
...
+                SELECT _error FROM test.kafka WHERE length(_error) != 0 ;
+        """
+        )
+
+        print(format_name)
+
+        kafka_create_topic(admin_client, f"{format_name}_err")
+
+        messages = ["qwertyuiop", "poiuytrewq", "zxcvbnm"]
+        kafka_produce(kafka_cluster, f'{format_name}_err', messages)
```
Comment (Avogar) [inline tests/integration/test_storage_kafka/test.py:4544]:
```suggestion
        kafka_produce(kafka_cluster, f"{format_name}_err", messages)
```

**Comment by Avogar** [comment]
Some really strage stuff is hapenning with kafka tests... Need to investigate more deeply

**Pull request merged**